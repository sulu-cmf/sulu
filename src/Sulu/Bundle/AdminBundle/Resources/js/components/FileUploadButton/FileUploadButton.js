// @flow
import React from 'react';
import Dropzone from 'react-dropzone';
import Button from '../Button';
import type {Node} from 'react';
import type {ButtonSkin} from '../Button';

type Props = {|
    accept?: string,
    children?: Node,
    disabled: boolean,
    icon?: string,
    onUpload: (file: File) => void,
    skin?: ButtonSkin,
|};

export default class FileUploadButton extends React.Component<Props> {
    static defaultProps = {
        accept: undefined,
        disabled: false,
        icon: undefined,
        skin: undefined,
    };

    handleDrop = (files: Array<File>) => {
        const file = files[0];

        this.props.onUpload(file);
    };

    render() {
        const {children, disabled, icon, skin, accept} = this.props;

        return (
            <Dropzone
                accept={accept ? {[accept]: []} : undefined}
                onDrop={this.handleDrop}
                style={{}}
            >
                {({getInputProps, getRootProps}) => (
                    <div {...getRootProps()}>
                        <Button disabled={disabled} icon={icon} skin={skin}>
                            {children}
                        </Button>
                        <input {...getInputProps()} />
                    </div>
                )}
            </Dropzone>
        );
    }
}
