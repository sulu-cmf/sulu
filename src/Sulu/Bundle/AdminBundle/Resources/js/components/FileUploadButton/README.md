The `FileUploadButton` is a button that opens a file dialog when it is clicked. After choosing a file the `onUpload`
callback is called with the file as its argument:

```javascript
const handleUpload = (file) => {
    alert('Uploaded file with ' + file.size + ' bytes');
};

<FileUploadButton onUpload={handleUpload}>
    Upload something
</FileUploadButton>
```

This component also accepts a `icon` and `skin` prop to allow customizing the displayed button:

```javascript
const handleUpload = (file) => {
    alert('Uploaded file with ' + file.size + ' bytes');
};

<FileUploadButton icon="su-image" onUpload={handleUpload} skin="link">
    Upload something
</FileUploadButton>
```
