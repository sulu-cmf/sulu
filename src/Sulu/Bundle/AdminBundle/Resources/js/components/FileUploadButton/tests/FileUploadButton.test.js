// @flow
import React from 'react';
import {mount, render} from 'enzyme';
import Dropzone from 'react-dropzone';
import FileUploadButton from '../FileUploadButton';

test('Render a FileUploadButton', () => {
    expect(render(<FileUploadButton onUpload={jest.fn()}>Upload something!</FileUploadButton>)).toMatchSnapshot();
});

test('Render a disabled FileUploadButton', () => {
    expect(render(
        <FileUploadButton disabled={true} onUpload={jest.fn()}>Upload something!</FileUploadButton>
    )).toMatchSnapshot();
});

test('Render a FileUploadButton with other skin and icon', () => {
    expect(render(
        <FileUploadButton icon="su-image" onUpload={jest.fn()} skin="link">Upload something!</FileUploadButton>
    )).toMatchSnapshot();
});

test('Call onUpload callback when a file is uploaded', () => {
    const uploadSpy = jest.fn();

    const fileUploadButton = mount(<FileUploadButton onUpload={uploadSpy}>Upload something!</FileUploadButton>);

    const testFileData = {name: 'test-file'};
    fileUploadButton.find(Dropzone).prop('onDrop')([testFileData]);

    expect(uploadSpy).toBeCalledTimes(1);
    expect(uploadSpy).toBeCalledWith(testFileData);
});
