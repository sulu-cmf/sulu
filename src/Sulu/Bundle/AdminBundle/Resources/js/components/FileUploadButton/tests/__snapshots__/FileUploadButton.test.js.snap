// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a FileUploadButton 1`] = `
<div>
  <div
    role="presentation"
    tabindex="0"
  >
    <button
      class="button secondary hasText"
      type="button"
    >
      <span
        class="buttonText"
      >
        Upload something!
      </span>
    </button>
    <input
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
  </div>
</div>
`;

exports[`Render a FileUploadButton with other skin and icon 1`] = `
<div>
  <div
    role="presentation"
    tabindex="0"
  >
    <button
      class="button link hasText"
      type="button"
    >
      <span
        aria-label="su-image"
        class="su-image buttonIcon"
      />
      <span
        class="buttonText"
      >
        Upload something!
      </span>
    </button>
    <input
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
  </div>
</div>
`;

exports[`Render a disabled FileUploadButton 1`] = `
<div>
  <div
    role="presentation"
    tabindex="0"
  >
    <button
      class="button secondary hasText"
      disabled=""
      type="button"
    >
      <span
        class="buttonText"
      >
        Upload something!
      </span>
    </button>
    <input
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
  </div>
</div>
`;
